File: src/data/courses/dsa/heaps/top-k/kth-smallest-element-in-sorted-matrix/code/python/kth_smallest_element_in_sorted_matrix.py
import heapq

class KthSmallestElementInSortedMatrix:
    def kthSmallest(self, matrix, k):
        n = len(matrix)
        min_heap = []
        
        for i in range(min(n, k)):
            heapq.heappush(min_heap, (matrix[i][0], i, 0))
        
        result = 0
        for _ in range(k):
            result, row, col = heapq.heappop(min_heap)
            
            if col + 1 < n:
                heapq.heappush(min_heap, (matrix[row][col + 1], row, col + 1))
        
        return result

File: src/data/courses/dsa/heaps/top-k/kth-smallest-element-in-sorted-matrix/code/java/KthSmallestElementInSortedMatrix.java
import java.util.*;

public class KthSmallestElementInSortedMatrix {
    public int kthSmallest(int[][] matrix, int k) {
        int n = matrix.length;
        PriorityQueue<int[]> minHeap = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        
        for (int i = 0; i < Math.min(n, k); i++) {
            minHeap.offer(new int[]{matrix[i][0], i, 0});
        }
        
        int result = 0;
        for (int i = 0; i < k; i++) {
            int[] curr = minHeap.poll();
            result = curr[0];
            int row = curr[1], col = curr[2];
            
            if (col + 1 < n) {
                minHeap.offer(new int[]{matrix[row][col + 1], row, col + 1});
            }
        }
        
        return result;
    }
}